### PR TITLE
add shvit model

### PR DESCRIPTION
**Pull request by learncat163**
[aistudio项目运行链接](https://aistudio.baidu.com/projectdetail/10696519)

# SHViT模型权重

## 1. 模型描述

SHViT（Single-Head Vision Transformer with Memory Efficient Macro Design, CVPR 2024）
采用内存高效的宏观结构设计：Patch embedding 由 4 级 stride-2 卷积构成，主体是 3 个
Stage，每个 Stage 由若干 `BasicBlock` 堆叠而成；`BasicBlock` 由 depthwise 卷积、
mixer、FFN 三个残差分支组成，其中 mixer 在 stage0 为 Identity、在后续 Stage 为单头
自注意力 `SHSA`；`SHSA` 只对部分通道（partial 通道）做 GroupNorm 后再生成 qkv，这是
单头与部分通道带来的内存节省来源；分类头为 `NormLinear`（BatchNorm1D + Linear）。
本工程在 PaddleClas 中增加了 SHViT_S1/S2/S3/S4 四个变体的支持。

### 2.1 整体权重概览

权重来源：

- **SHViT**: 本工程的主要目的，从 timm 权重转换过来到 paddle 项目

**SHViT（timm 权重转换）**

| PaddleClas 模型 | 原始timm 模型                 | paddle转换权重         | 文件大小 |
| --------------- | ----------------------------- | ---------------------- | -------- |
| SHViT_S1        | `timm/shvit_s1.in1k`          | `SHViT_S1.pdparams`    | 25.53M   |
| SHViT_S2        | `timm/shvit_s2.in1k`          | `SHViT_S2.pdparams`    | 46.20M   |
| SHViT_S3        | `timm/shvit_s3.in1k`          | `SHViT_S3.pdparams`    | 57.29M   |
| SHViT_S4        | `timm/shvit_s4.in1k`          | `SHViT_S4.pdparams`    | 66.73M   |

各变体规格：

| 变体     | 输入尺寸 | 可训练参数量 | timm 配置                                                                    |
| -------- | -------- | ------------ | ---------------------------------------------------------------------------- |
| SHViT_S1 | 224      | 6.33 M       | embed_dim=(128,224,320) depth=(2,4,5) partial_dim=(32,48,68) types=("i","s","s") |
| SHViT_S2 | 224      | 11.48 M      | embed_dim=(128,308,448) depth=(2,4,5) partial_dim=(32,66,96) types=("i","s","s") |
| SHViT_S3 | 224      | 14.25 M      | embed_dim=(192,352,448) depth=(3,5,5) partial_dim=(48,75,96) types=("i","s","s") |
| SHViT_S4 | 256      | 16.59 M      | embed_dim=(224,336,448) depth=(4,7,6) partial_dim=(48,72,96) types=("i","s","s") |

### 2.2 权重获取说明

**SHViT** - 先下载 timm 权重，再转换为 Paddle 格式。

转换过程中的关键处理：上游权重与 PaddleClas 的差异只集中在三处——torch BN 的训练
计数需要丢弃、BN 统计量的 key 需要改名、分类头 Linear 的权重需要转置；卷积、GroupNorm
与 BatchNorm1D 的张量布局两端一致，原样直通。

使用转换脚本 `convert_weights.py` 进行权重转换，脚本与本文档同目录。

### 3.1 权重转换的逐 key 对齐

因为 SHViT 的权重全部由上游 timm 权重转换得到，所以必须保证转换过程本身不引入任何
数值改动；否则后面的前向对齐与 ImageNet 精度验证都没有意义。

转换后的 pdparams 需要与 PaddleClas 模型的 `state_dict()` 一一对应，不能出现多余的
key、缺失的 key 或者形状不匹配的 key（Paddle 的 `set_dict` 对形状不匹配是静默跳过，
不会报错，因此不能依赖框架的容错）。

所以对齐要求是：把上游 safetensors 中的每一个张量，与转换后 pdparams 中对应 key 的
张量逐元素比对，要求完全相等。

#### 3.1.1 权重逐 key 比对结果

| 变体     | key 数 | 最大绝对误差 | 逐 key 全等 | 结论 |
| -------- | ------ | ------------ | ----------- | ---- |
| SHViT_S1 | 397    | 0.0          | True        | PASS |
| SHViT_S2 | 397    | 0.0          | True        | PASS |
| SHViT_S3 | 439    | 0.0          | True        | PASS |
| SHViT_S4 | 535    | 0.0          | True        | PASS |

所有变体的权重逐 key 相等。

#### 3.1.2 转换与校验的处理逻辑

```python
"""按同一个映射把上游张量还原，与转换结果逐 key 比对。"""

for name, upstream in VARIANTS.items():
    raw = load_upstream_weights(upstream)          # timm safetensors
    converted = load_pdparams(name)                # 转换产出的 pdparams

    # 转换只做 key 改名 / 转置 / 丢弃, 逐 key 还原回去后必须完全相等
    restored = restore_to_upstream_layout(converted)

    assert keys_match(restored, raw), "key 覆盖不一致"
    assert all_equal(restored, raw), "存在数值差异, 转换有损"
```

已经经过验证，四个变体的转换权重与上游权重逐 key 完全一致，最大绝对误差为 0。

### 3.2 前向推理精度对齐

通过加载转换后的 Paddle 权重，与 PyTorch timm 原版模型进行前向推理对比（两侧在同一设备 GPU 上执行、torch 侧关闭 TF32、使用同一份固定随机种子的输入，判定阈值 `max_abs_diff < 1e-4`）：

**SHViT**

| 模型     | 输入尺寸 | 最大绝对误差 | 平均绝对误差 | Top-1 一致 |
| -------- | -------- | ------------ | ------------ | ---------- |
| SHViT_S1 | 224x224  | 1.550e-06    | 2.533e-07    | True       |
| SHViT_S2 | 224x224  | 9.909e-07    | 2.591e-07    | True       |
| SHViT_S3 | 224x224  | 1.058e-06    | 2.382e-07    | True       |
| SHViT_S4 | 256x256  | 7.749e-07    | 1.955e-07    | True       |

### 3.3 ImageNet 验证集精度验证

在 ImageNet 验证集上验证模型的 Top-1 准确率：

**SHViT**

| 模型     | 输入尺寸 | Paddle Acc | Timm Acc | 误差  |
| -------- | -------- | ---------- | -------- | ----- |
| SHViT_S1 | 224x224  | 72.78%     | 72.78%   | 0.00% |
| SHViT_S2 | 224x224  | 75.18%     | 75.18%   | 0.00% |
| SHViT_S3 | 224x224  | 77.36%     | 77.36%   | 0.00% |
| SHViT_S4 | 256x256  | 79.36%     | 79.36%   | 0.00% |

所有模型的 Top-1 准确率与 timm 版本完全一致，误差为 0.00%，满足精度要求。验证在
ImageNet 验证集全量 50000 张上进行，以 timm 预测为参考标签做严格 Top-1 比对，四个
变体一致率均为 100.0000%，因此 PaddleClas 侧准确率与 timm 侧严格相等；Timm Acc 一列
取自 timm 仓库的 `results/results-imagenet.csv`。

### CODE REVIEW

**Comment by learncat163** [comment]
# 复现方式

固定 batch（4 张随机图 + 固定标签）反复训练 50 步，验证模型能对该 batch 拟合
（loss 下降）。Optimizer 为 AdamW（lr=1e-4, weight_decay=0.05），Loss 为
CrossEntropyLoss，加载转换后的预训练权重。


# 结果

### SHViT_S1 (224x224) 逐步 Loss

| Step | Loss |
|------|------|
| 0 | 7.7036 |
| 1 | 5.5728 |
| 2 | 4.1893 |
| 3 | 2.9773 |
| 4 | 2.0048 |
| 5 | 1.2880 |
| 6 | 0.9020 |
| 7 | 0.5877 |
| 8 | 0.3827 |
| 9 | 0.2682 |
| 10 | 0.2027 |
| 11 | 0.1665 |
| 12 | 0.1430 |
| 13 | 0.1278 |
| 14 | 0.1131 |
| 15 | 0.0991 |
| 16 | 0.0878 |
| 17 | 0.0791 |
| 18 | 0.0717 |
| 19 | 0.0650 |
| 20 | 0.0589 |
| 21 | 0.0537 |
| 22 | 0.0493 |
| 23 | 0.0459 |
| 24 | 0.0429 |
| 25 | 0.0403 |
| 26 | 0.0378 |
| 27 | 0.0355 |
| 28 | 0.0335 |
| 29 | 0.0317 |
| 30 | 0.0301 |
| 31 | 0.0287 |
| 32 | 0.0275 |
| 33 | 0.0264 |
| 34 | 0.0254 |
| 35 | 0.0245 |
| 36 | 0.0236 |
| 37 | 0.0228 |
| 38 | 0.0220 |
| 39 | 0.0212 |
| 40 | 0.0205 |
| 41 | 0.0199 |
| 42 | 0.0193 |
| 43 | 0.0188 |
| 44 | 0.0183 |
| 45 | 0.0178 |
| 46 | 0.0173 |
| 47 | 0.0169 |
| 48 | 0.0165 |
| 49 | 0.0162 |

### SHViT_S2 (224x224) 逐步 Loss

| Step | Loss |
|------|------|
| 0 | 7.4963 |
| 1 | 6.5336 |
| 2 | 5.0340 |
| 3 | 3.8178 |
| 4 | 2.3908 |
| 5 | 1.5261 |
| 6 | 1.1157 |
| 7 | 0.5850 |
| 8 | 0.3694 |
| 9 | 0.2638 |
| 10 | 0.1892 |
| 11 | 0.1492 |
| 12 | 0.1209 |
| 13 | 0.0988 |
| 14 | 0.0817 |
| 15 | 0.0687 |
| 16 | 0.0587 |
| 17 | 0.0507 |
| 18 | 0.0446 |
| 19 | 0.0395 |
| 20 | 0.0354 |
| 21 | 0.0319 |
| 22 | 0.0290 |
| 23 | 0.0266 |
| 24 | 0.0244 |
| 25 | 0.0226 |
| 26 | 0.0210 |
| 27 | 0.0196 |
| 28 | 0.0184 |
| 29 | 0.0173 |
| 30 | 0.0164 |
| 31 | 0.0155 |
| 32 | 0.0147 |
| 33 | 0.0140 |
| 34 | 0.0133 |
| 35 | 0.0127 |
| 36 | 0.0122 |
| 37 | 0.0117 |
| 38 | 0.0113 |
| 39 | 0.0109 |
| 40 | 0.0106 |
| 41 | 0.0103 |
| 42 | 0.0100 |
| 43 | 0.0097 |
| 44 | 0.0095 |
| 45 | 0.0092 |
| 46 | 0.0090 |
| 47 | 0.0088 |
| 48 | 0.0086 |
| 49 | 0.0084 |

### SHViT_S3 (224x224) 逐步 Loss

| Step | Loss |
|------|------|
| 0 | 6.7551 |
| 1 | 6.0289 |
| 2 | 4.5494 |
| 3 | 2.6638 |
| 4 | 1.9072 |
| 5 | 1.0441 |
| 6 | 0.5802 |
| 7 | 0.3452 |
| 8 | 0.2447 |
| 9 | 0.1836 |
| 10 | 0.1350 |
| 11 | 0.1037 |
| 12 | 0.0829 |
| 13 | 0.0680 |
| 14 | 0.0577 |
| 15 | 0.0501 |
| 16 | 0.0440 |
| 17 | 0.0393 |
| 18 | 0.0353 |
| 19 | 0.0320 |
| 20 | 0.0291 |
| 21 | 0.0266 |
| 22 | 0.0244 |
| 23 | 0.0225 |
| 24 | 0.0209 |
| 25 | 0.0195 |
| 26 | 0.0182 |
| 27 | 0.0171 |
| 28 | 0.0160 |
| 29 | 0.0151 |
| 30 | 0.0143 |
| 31 | 0.0136 |
| 32 | 0.0130 |
| 33 | 0.0124 |
| 34 | 0.0119 |
| 35 | 0.0114 |
| 36 | 0.0110 |
| 37 | 0.0106 |
| 38 | 0.0102 |
| 39 | 0.0099 |
| 40 | 0.0095 |
| 41 | 0.0092 |
| 42 | 0.0089 |
| 43 | 0.0087 |
| 44 | 0.0084 |
| 45 | 0.0082 |
| 46 | 0.0080 |
| 47 | 0.0078 |
| 48 | 0.0076 |
| 49 | 0.0075 |

### SHViT_S4 (256x256) 逐步 Loss

| Step | Loss |
|------|------|
| 0 | 6.7162 |
| 1 | 7.3408 |
| 2 | 6.6833 |
| 3 | 6.0761 |
| 4 | 4.8228 |
| 5 | 4.1792 |
| 6 | 3.5411 |
| 7 | 2.5341 |
| 8 | 2.2929 |
| 9 | 1.4898 |
| 10 | 1.1034 |
| 11 | 0.8450 |
| 12 | 0.6310 |
| 13 | 0.4118 |
| 14 | 0.2793 |
| 15 | 0.1911 |
| 16 | 0.1449 |
| 17 | 0.1231 |
| 18 | 0.1034 |
| 19 | 0.0851 |
| 20 | 0.0701 |
| 21 | 0.0590 |
| 22 | 0.0505 |
| 23 | 0.0442 |
| 24 | 0.0397 |
| 25 | 0.0364 |
| 26 | 0.0336 |
| 27 | 0.0312 |
| 28 | 0.0291 |
| 29 | 0.0271 |
| 30 | 0.0253 |
| 31 | 0.0238 |
| 32 | 0.0225 |
| 33 | 0.0214 |
| 34 | 0.0203 |
| 35 | 0.0193 |
| 36 | 0.0184 |
| 37 | 0.0175 |
| 38 | 0.0168 |
| 39 | 0.0161 |
| 40 | 0.0155 |
| 41 | 0.0149 |
| 42 | 0.0144 |
| 43 | 0.0139 |
| 44 | 0.0135 |
| 45 | 0.0131 |
| 46 | 0.0127 |
| 47 | 0.0124 |
| 48 | 0.0120 |
| 49 | 0.0117 |

## 汇总

| 变体 | 输入 | Step 0 | Step 49 | 是否单调下降 | 耗时 |
|------|------|--------|---------|--------------|------|
| SHViT_S1 | 224 | 7.7036 | 0.0162 | 是 | 2.1 s |
| SHViT_S2 | 224 | 7.4963 | 0.0084 | 是 | 1.6 s |
| SHViT_S3 | 224 | 6.7551 | 0.0075 | 是 | 2.2 s |
| SHViT_S4 | 256 | 6.7162 | 0.0117 | 否（仅 step 0->1 抬升） | 2.3 s |
